drivers/bridge/bridge.go

```diff
@@ -1451,3 +1479,60 @@ func electMacAddress(epConfig *endpointConfiguration, ip net.IP) net.HardwareAdd
 	}
 	return netutils.GenerateMACFromIP(ip)
 }
+
+func (d *driver) RestoreBridgeEndpoint(n *bridgeNetwork, endpoint *bridgeEndpoint, iface driverapi.InterfaceInfo) error {
+	ipv4Addr := iface.Address()
+	// Restore ipv4
+	_, err := ipAllocator.RequestIP(n.bridge.bridgeIPv4, ipv4Addr.IP)
+	if err != nil {
+		return err
+	}
+
+	n.Lock()
+	config := n.config
+	n.Unlock()
+
+	// Restore ipv6
+	ipv6Addr := iface.AddressIPv6()
+	if ipv6Addr.IP != nil && len(ipv6Addr.IP) != 0 {
+		network := n.bridge.bridgeIPv6
+		if config.FixedCIDRv6 != nil {
+			network = config.FixedCIDRv6
+		}
+		_, err := ipAllocator.RequestIP(network, ipv6Addr.IP)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Restore container interface name
+	sbox, err := linkByMacAddress(iface.MacAddress().String())
+	if err != nil {
+		return err
```

Comment:
I tested it using unit tests in the code mainly. I have tested aginst docker before I introduced restoring portmapping/macadress. I apologize for the errors you encounted. I should be more careful. For the user case, I have nothing else except for hot upgrade. And it seems you have something depending on localstore, so I suggest to split the last commit.